File: backend/internal/model/calculator.go
package model

type ProcessStep struct {
	Operation      string  `json:"operation"`
	Executor       string  `json:"executor,omitempty"`
	MinutesPerUnit float64 `json:"minutes_per_unit"`
	RateRubPerMin  float64 `json:"rate_rub_per_min"`
}

type CalculatorInput struct {
	ProcessName string `json:"process_name"`

	// Step 1: process assessment
	HmMode          string        `json:"hm_mode"` // "process" | "direct"
	ProcessSteps    []ProcessStep `json:"process_steps,omitempty"`
	UnitsPerMonth   float64       `json:"units_per_month"`
	AutomationPct   float64       `json:"automation_pct"`
	HoursSavedMonth float64       `json:"hours_saved_month"` // Hm — direct or overridden

	// Optional KPI helpers
	ErrorRateBeforePct    float64 `json:"error_rate_before_pct"`
	ThroughputBeforePerDay float64 `json:"throughput_before_per_day"`

	// Step 2: financial model (Excel)
	HourlyCostLoaded     float64 `json:"hourly_cost_loaded"`      // Ch
	Utilization          float64 `json:"utilization"`             // L
	OtherBenefitMonthly  float64 `json:"other_benefit_monthly"`   // Sm
	MonthlySolutionCost  float64 `json:"monthly_solution_cost"`   // Om
	Capex                float64 `json:"capex"`                   // I0
	HorizonMonths        int     `json:"horizon_months"`          // n
	DiscountRateAnnual   float64 `json:"discount_rate_annual"`    // r
}

type Recommendation string

const (
	RecommendationAutomate       Recommendation = "automate"
	RecommendationConsider       Recommendation = "consider"
	RecommendationNotRecommended Recommendation = "not_recommended"
)

type KPIRow struct {
	Key    string `json:"key"`
	Label  string `json:"label"`
	Before string `json:"before"`
	After  string `json:"after"`
	Change string `json:"change"`
}

type CalculatorOutput struct {
	HoursSavedMonth   float64 `json:"hours_saved_month"`
	MinutesPerUnit    float64 `json:"minutes_per_unit"`
	NetBenefitMonthly float64 `json:"net_benefit_monthly"` // B_m
	FTE               float64 `json:"fte"`
	PaybackMonths     float64 `json:"payback_months"`
	ROIHorizonPct     float64 `json:"roi_horizon_pct"`
	TCOHorizon        float64 `json:"tco_horizon"`
	NPV               float64 `json:"npv"`
	PayrollSavings    float64 `json:"payroll_savings"`
	MonthlyHoursBefore float64 `json:"monthly_hours_before"`
	MonthlyHoursAfter  float64 `json:"monthly_hours_after"`

	Recommendation     Recommendation `json:"recommendation"`
	RecommendationText string         `json:"recommendation_text"`
	KPIRows            []KPIRow         `json:"kpi_rows"`
}

File: frontend/src/lib/calculator.ts
import type { CalculatorInput, CalculatorOutput, ProcessStep } from "./api";

export function totalMinutesPerUnit(steps: ProcessStep[]): number {
  return steps.reduce((s, r) => s + (r.minutes_per_unit || 0), 0);
}

export function resolveHoursSaved(input: CalculatorInput): number {
  if (input.hm_mode === "direct") {
    return input.hours_saved_month;
  }
  const totalMin = totalMinutesPerUnit(input.process_steps || []);
  if (totalMin <= 0 || input.units_per_month <= 0 || input.automation_pct <= 0) {
    return input.hours_saved_month;
  }
  return (totalMin * input.units_per_month * input.automation_pct) / 100 / 60;
}

export function previewCalculator(
  input: CalculatorInput,
  locale = "ru"
): CalculatorOutput {
  const Hm = resolveHoursSaved(input);
  const L = input.utilization > 0 && input.utilization <= 1 ? input.utilization : 0.8;
  const Ch = input.hourly_cost_loaded;
  const Sm = input.other_benefit_monthly;
  const Om = input.monthly_solution_cost;
  const I0 = input.capex;
  const n = input.horizon_months > 0 ? input.horizon_months : 12;
  const r = input.discount_rate_annual > 0 ? input.discount_rate_annual : 0.2;

  const Bm = Hm * Ch + Sm - Om;
  const FTE = L > 0 ? Hm / (168 * L) : 0;
  const payback = Bm > 0 ? I0 / Bm : Infinity;
  const roiPct = I0 > 0 ? ((Bm * n - I0) / I0) * 100 : 0;
  const tco = I0 + Om * n;

  const rm = Math.pow(1 + r, 1 / 12) - 1;
  let npv = -I0;
  for (let t = 1; t <= n; t++) {
    npv += Bm / Math.pow(1 + rm, t);
  }

  let rec: CalculatorOutput["recommendation"] = "automate";
  let text =
    locale === "en"
      ? "Automation is recommended"
      : "Рекомендуется автоматизировать";
  if (Bm <= 0 || !isFinite(payback) || payback > 24) {
    rec = "not_recommended";
    text =
      locale === "en"
        ? "Automation is not recommended"
        : "Автоматизация не рекомендуется";
  } else if (payback >= 12 || roiPct <= 0) {
    rec = "consider";
    text =
      locale === "en" ? "Consider automation" : "Рассмотрите автоматизацию";
  }

  return {
    hours_saved_month: Hm,
    minutes_per_unit: totalMinutesPerUnit(input.process_steps || []),
    net_benefit_monthly: Bm,
    fte: FTE,
    payback_months: payback,
    roi_horizon_pct: roiPct,
    tco_horizon: tco,
    npv,
    payroll_savings: Hm * Ch,
    monthly_hours_before: 0,
    monthly_hours_after: 0,
    recommendation: rec,
    recommendation_text: text,
    kpi_rows: [],
  };
}

export const defaultCalculatorInput: CalculatorInput = {
  process_name: "",
  hm_mode: "process",
  process_steps: [
    {
      operation: "Ввод счета в систему",
      executor: "Оператор",
      minutes_per_unit: 8,
      rate_rub_per_min: 17,
    },
    {
      operation: "Проверка и подтверждение",
      executor: "Менеджер",
      minutes_per_unit: 2,
      rate_rub_per_min: 33,
    },
  ],
  units_per_month: 8000,
  automation_pct: 80,
  hours_saved_month: 800,
  error_rate_before_pct: 5,
  throughput_before_per_day: 20,
  hourly_cost_loaded: 2500,
  utilization: 0.8,
  other_benefit_monthly: 15000,
  monthly_solution_cost: 10000,
  capex: 300000,
  horizon_months: 12,
  discount_rate_annual: 0.2,
};
